glfw-test.rkt

```diff
@@ -1,58 +1,60 @@
 #lang racket/base
-(require (planet schematics/schemeunit:3:4))
-(require (planet schematics/schemeunit:3:4/text-ui))
-(require "glfw.rkt")
-
-(define-test-suite glfw-init
-  #:before (lambda () (glfwInit))
-  #:after (lambda () (glfwTerminate)) 
-
-  (test-case "Version query succeeds with non-zero values"
-             (let-values ([(major minor rev) (glfwGetVersion)]) 
-               (check-not-eq? major 0)
-               (check-not-eq? minor 0)
-               (check-not-eq? rev 0)))
-
-  (test-case "Opening window in windowed mode succeeds"
-             (glfwOpenWindow 400 300
-                             8 8 8
-                             0 0 0
-                             GLFW_WINDOW)
-             (glfwCloseWindow))
-
-  (test-case "Opening window in fullscreen mode succeeds"
-             (glfwOpenWindow 640 480
-                             8 8 8
-                             0 0 0
-                             GLFW_FULLSCREEN)
-             (glfwCloseWindow)))
-
-(define-test-suite glfw-window-functions
-  #:before (lambda ()
-             (glfwInit)
-             (glfwOpenWindow 400 300
-                             8 8 8
-                             0 0 0
-                             GLFW_WINDOW))
-  #:after (lambda ()
-            (glfwCloseWindow)
-            (glfwTerminate))
-
-  (test-case "Setting and getting window size succeeds"
-             (glfwSetWindowSize 350 250)
-             (let-values ([(x y) (glfwGetWindowSize)])
-               (check-eq? x 350)
-               (check-eq? y 250)))
-
-  (test-case "Setting window position succeeds"
-             (glfwSetWindowPos 10 10))
-
-  (test-case "Setting window title succeeds"
-             (glfwSetWindowTitle "Test title"))
-
-  (test-case "Window iconify and restore succeeds"
-             (glfwIconifyWindow)
-             (glfwRestoreWindow)))
-
-(run-tests glfw-init)
-(run-tests glfw-window-functions)
+
+(module+ test
+  (require "glfw.rkt"
+           disposable
+           fixture
+           racket/function
+           rackunit
+           rackunit/text-ui)
+
+  (define (check-not-= x y) (check-false (= x y)))
+
+  (define-test-suite initialization
+    #:before glfwInit
+    #:after glfwTerminate
+
+    (test-case "Version query succeeds with non-zero values"
+               (let-values ([(major minor rev) (glfwGetVersion)])
+                 (check-not-= major 0)
+                 (check-not-= minor 0)
+                 (check-not-= rev 0)))
+
+    (test-case "Opening window in windowed mode succeeds"
+               (let ([window (glfwCreateWindow 400 300 "Main window" #f #f)])
+                 (glfwDestroyWindow window)))
+
+    (test-case "Opening window in fullscreen mode succeeds"
+               (let ([window (glfwCreateWindow 400 300 "Main window" (glfwGetPrimaryMonitor) #f)])
+                 (glfwDestroyWindow window))))
+
+  (define (create-window)
+    (glfwInit)
```

Comment:
Do bad things happen if `glfwInit` is called twice before `glfwTerminate` is called? Or if terminate is called twice in a row? If so, it might be best not to include creating and destroying it in the window disposable. The fixture library will make one window for each test case *and* one extra window for the whole group of tests, which doesn't work so great for global setup and teardown.

I think you could remove the init and teardown and instead move the `"window-functions"` test cases inside the `initializing` test suite. That way they reuse that suite's init and teardown logic and don't need their own.

Reply:
Why does it create one for the whole group too? No way to prevent this?

Reply:
Sometimes people put both checks and sub-tests inside a `test-case` like this:

```
(test-case "group"
  (check-sanity-or-bail-on-all-tests-early)
  (test-case "case1" ...)
  (test-case "case2" ...))
```

So to avoid surprises, `text-case/fixture` makes sure those tests have access to fixtures as well. But they shouldn't have the *same* fixture instance as any nested tests, since fixtures are supposed to provided isolated access to resources.

But yeah, this makes `glfwInit` and `glfwTerminate` do surprising things inside a fixture. Disposables in general aren't great with global singletons; they're primarily designed for resources you could make multiples of. That could probably be improved but I'm not sure how yet.